### PR TITLE
[lv_list] Add list layout, fix #1138

### LINKED ISSUE #1138: Horizontal list problem
#### 1. One button is fine
image

```c++
void App::create_navigation_bar(lv_obj_t *parent) {
    lv_obj_t *list = lv_list_create(parent, NULL);
    lv_cont_set_fit2(list, LV_FIT_FLOOD, LV_FIT_TIGHT);
    lv_page_set_scrl_layout(list, LV_LAYOUT_ROW_T);
    lv_page_set_scrl_fit(list, LV_FIT_TIGHT);
    
    lv_list_add_btn(list, NULL, "button1");
}
```

---

#### 2.Two button, the button becomes very long
image
```c++
void App::create_navigation_bar(lv_obj_t *parent) {
    lv_obj_t *list = lv_list_create(parent, NULL);
    lv_cont_set_fit2(list, LV_FIT_FLOOD, LV_FIT_TIGHT);
    lv_page_set_scrl_layout(list, LV_LAYOUT_ROW_T);
    lv_page_set_scrl_fit(list, LV_FIT_TIGHT);

    lv_list_add_btn(list, NULL, "button1");
    lv_list_add_btn(list, NULL, "button1");
}

```
---

#### 3.And try to add other elements to list will cause exception 
image

```c++
void App::create_navigation_bar(lv_obj_t *parent) {
    lv_obj_t *list = lv_list_create(parent, NULL);
    lv_cont_set_fit2(list, LV_FIT_FLOOD, LV_FIT_TIGHT);
    lv_page_set_scrl_layout(list, LV_LAYOUT_ROW_T);
    lv_page_set_scrl_fit(list, LV_FIT_TIGHT);

    lv_list_add_btn(list, NULL, "button1");
    lv_list_add_btn(list, NULL, "button1");

    lv_label_set_text(lv_label_create(list, NULL), "Test add other element");
}
```

---

### 4. But I change this line https://github.com/littlevgl/lvgl/blob/8e9764532d866719b5a87f0e2181f7d57c501592/src/lv_objx/lv_list.c#L196 to
```c
lv_btn_set_fit2(liste, LV_FIT_TIGHT, LV_FIT_TIGHT)
``` 
### All the problemes were gone
image

### PR DESCRIPTION
**Pull request by C47D**
Hi,

This is to implement the list layout, let me know if the prototypes are right before implementing the functions.

Carlos

### CODE REVIEW

File: src/lv_objx/lv_list.h [hunk 1]
```diff
@@ -78,6 +78,13 @@ enum {
 };
 typedef uint8_t lv_list_style_t;
 
+/** List layouts. **/
```
Comment (embeddedt) [inline src/lv_objx/lv_list.h]: I'm being really picky with this, but the end comment can be `*/` not `**/`.

Comment (C47D) [reply src/lv_objx/lv_list.h]: Will fix it.

**Comment by C47D** [comment]
@embeddedt should we add a layout bitfield in the list struct?

**Comment by embeddedt** [comment]
@C47D Yes. No need to document it, as it is an internal implementation.

**Comment by C47D** [comment]
@embeddedt what layout should be the default layout?

Also, documentation for `lv_list_get_prev_btn` and `lv_list_get_next_btn` are backwards.
```
/**
 * Get the next button from list. (Starts from the bottom button)
 * @param list pointer to a list object
 * @param prev_btn pointer to button. Search the next after it.
 * @return pointer to the next button or NULL when no more buttons
 */
lv_obj_t * lv_list_get_prev_btn(const lv_obj_t * list, lv_obj_t * prev_btn);

/**
 * Get the previous button from list. (Starts from the top button)
 * @param list pointer to a list object
 * @param prev_btn pointer to button. Search the previous before it.
 * @return pointer to the previous button or NULL when no more buttons
 */
lv_obj_t * lv_list_get_next_btn(const lv_obj_t * list, lv_obj_t * prev_btn);

```

Do you want me to fix them in this PR? If so, how should them be documented?

**Comment by C47D** [comment]
@embeddedt please let me know if the implementation makes sense, i'm not very familiar with the lvgl source. :+1:

**Comment by embeddedt** [comment]
Will review implementation in a few hours. Vertical layout should be default for backwards compatibility.

File: src/lv_objx/lv_list.c [hunk 2]
```diff
@@ -366,6 +367,36 @@ void lv_list_set_style(lv_obj_t * list, lv_list_style_t type, const lv_style_t *
...
+        while(btn != NULL) {
+			if (LV_LIST_LAYOUT_HOR == layout) {
+				lv_btn_set_fit2(list, LV_FIT_FLOOD, LV_FIT_TIGHT);
+			} else { /* LV_LIST_LAYOUT_VER */
+				lv_btn_set_fit(list, LV_FIT_TIGHT);
+			}
+
+            btn = lv_list_get_prev_btn(list, btn);
+        }
+
```
Comment (kisvegabor) [inline src/lv_objx/lv_list.c]: It seems you forget to apply the layout with `lv_page_set_scr_layout(list, LV_LAYOUT_COL_M/ROW_M)`

Comment (kisvegabor) [inline src/lv_objx/lv_list.c:392]: It seems you forget to apply the layout with lv_page_set_scr_layout(list, LV_LAYOUT_COL_M/ROW_M)

**Review by kisvegabor — COMMENTED** [review]
Look good thank you! I've only one comment.

I'm thinking about if `lv_list_layout_t` is really required. It would give more flexibility to use `lv_layout_t` with `LV_LAYOUT_COL_M` and `LV_LAYOUT_ROW_T/B/M`. Or finally when we added all list elements and all positions are set the layout can be changed to `LV_LAYOUT_OFF` to apply an animation (like floating in the elements.)

What do you think?

**Comment by C47D** [comment]
Hi,

Thanks for the review @kisvegabor, i am almost 0% familiar with lvgl source and internal working, please let me know if b4b4c76 solves the issues you found.

In regards of LV_LAYOUT_OFF i didn't really understood, is that at the end of the creation of the list?
I don't have a lot of free time from Monday to Friday (because of work), so if you want to take the PR from here and add the missing pieces please do so.

Carlos

File: src/lv_objx/lv_list.c (hunk 2 above)
Comment (C47D) [reply src/lv_objx/lv_list.c:392]: Does this `lv_page_set_scr_layout(list, layout == LV_LAYOUT_COL_M ? LV_LAYOUT_COL_M : LV_LAYOUT_ROW_M);` looks fine?

**Comment by kisvegabor** [comment]
> In regards of LV_LAYOUT_OFF i didn't really understood, is that at the end of the creation of the list?

Let's say you create a list with e.g. `COL_M` layout and you want to float in the buttons after each other you need to disable the layout to enable positioning via animations. 

I've pushed an update  a568a13
It looks good to me. If you agree I can merge it.

**Comment by C47D** [comment]
> It looks good to me. If you agree I can merge it.

It looks good to me. I haven't tested it on hardware/simulator tho.

I think in 5f9324c you merged my branch somewhere already.

> Let's say you create a list with e.g. COL_M layout and you want to float in the buttons after each other you need to disable the layout to enable positioning via animations.

Do you have any suggestion on how to implement this?

**Comment by kisvegabor** [comment]
> I think in 5f9324c you merged my branch somewhere already.

When I've switched from master to your feature brach there was merge with the updates in master.


> Do you have any suggestion on how to implement this?

Sure, here is an example:

list

```c
   lv_obj_t * list = lv_list_create(lv_scr_act(), NULL);
    lv_list_add_btn(list, NULL, "First");
    lv_list_add_btn(list, NULL, "Second");
    lv_list_add_btn(list, NULL, "Third");
    lv_list_add_btn(list, NULL, "Forth");
    lv_list_set_layout(list, LV_LAYOUT_OFF);

    lv_obj_t * b;
    b = lv_list_get_next_btn(list, NULL);
    lv_anim_t  a_pos;
    lv_anim_init(&a_pos);
    lv_anim_set_values(&a_pos, 30, 5);

    lv_anim_t  a_opa;
    lv_anim_init(&a_opa);
    lv_anim_set_values(&a_opa, LV_OPA_TRANSP, LV_OPA_COVER);

    uint32_t delay = 1000;
    while(b!= NULL) {

        lv_anim_set_exec_cb(&a_pos, b, lv_obj_set_x);
        lv_anim_set_time(&a_pos, 300, delay);
        lv_anim_create(&a_pos);


        lv_obj_set_opa_scale_enable(b, true);
        lv_anim_set_exec_cb(&a_opa, b, lv_obj_set_opa_scale);
        lv_anim_set_time(&a_opa, 300, delay);
        lv_anim_create(&a_opa);

        b = lv_list_get_next_btn(list, b);
        delay += 100;
    }

    lv_obj_set_x(lv_list_get_next_btn(list, NULL), 20);
```

**Comment by C47D** [comment]
Can we merge and close this PR before adding the support LV_LAYOUT_OFF? If i understand correctly the layout before the animation should be cached and restored after the animation is done, isn't?

**Comment by kisvegabor** [comment]
As you can see from the example LAYOUT_OFF is already working. We don't need to cache the previous layout because it's much easier for the user to call `lv_list_set_layout` again when the animations (or anything) is ready.

I've just changed the base `dev-6.1` and merged the PR.

Thank you very much!

**Pull request merged**